### PR TITLE
masked_iterate, masked_iterate_final (GEN-327, GEN-859)

### PR DESCRIPTION
**Pull request by sritchie**
This PR:

- adds `genjax.masked_iterate` and `genjax.masked_iterate_final`
- adds `masked_iterate` and `masked_iterate_final` combinator methods to `GenerativeFunction`
- adds new docs
- modifies the masking notebook to use this combinator, instead of the previous inlined version
- also fixes a bug where `genjax.flip` returned an int instead of a boolean

### CODE REVIEW

**Review by MathieuHuot — APPROVED** [review]
Looks great, excited to see more combinators :).
